### ISSUE TITLE
Graceful shutdown of server

### ISSUE THREAD

**Issue by muja** [iron/iron#357]
For purposes outside of webdev, it is sometimes useful to listen for a specific request and shutdown the server afterwards (while still allowing to respond). This current isn't possible with iron, and hopefully isn't out of the project's scope. I would love to see it implemented! Code snippet of how it might look:

```rust
let iron = Iron::new(hello_world);
println!("Start listening!");
iron.http("localhost:3000");
println!("Done listening!");

fn hello_world(_: &mut Request) -> IronResult<Response> {
    iron.graceful_shutdown()
    Ok(Response::with((iron::status::Ok, "Hello World")))
}
```

**Comment by reem** [comment]
@muja I agree completely, this is functionality we are looking to develop in hyper which we could then expose in iron.

**Comment by muja** [comment]
Thanks for the response! So hyper doesn't support this either, good to know.

Is this feature already envisioned / when should we expect it?

**Comment by reem** [comment]
@muja currently the std library doesn't provide a way to close a `TcpListener`. As soon as it does, we will add support in hyper and then I can add support in iron.

**Comment by muja** [comment]
What about `drop(listener)`?

From the [docs](https://doc.rust-lang.org/std/net/struct.TcpListener.html):
```
let listener = TcpListener::bind("127.0.0.1:80").unwrap();

[...]

// close the socket server
drop(listener);
```

**Comment by Gyscos** [comment]
I think he meant once `accept()` is called, it blocks until a connection is made and cannot be cancelled.

**Comment by jmatraszek** [comment]
Hi @gyscos,
could you please share a snippet with working workaround for this? I'll trying to implement this using dummy connection, but I fail to close Iron even after that. I would really appreciate that.

**Comment by jmatraszek** [comment]
Hi, any update on this? AFAIK it is not possible to close the listener in hyper now (at least in 0.11).

**Comment by kargakis** [comment]
FWIW, it seems hyper can do a graceful shutdown in 0.11

https://github.com/hyperium/hyper/issues/338#issuecomment-347044047
